refactor fourth order tensor fill methods for efficiency

lynnmunday opened the pull request:
-This hardcodes the filling of elastic materials properties fourth order tensors
-The fill method was also split into two methods depending on the type of anisotropy
-isotropic materials call the new method to  fill there tensors.
-unit test for the two types of anisotropic tensors
ref #15404

lynnmunday commented:
@tophmatthews @dschwen @recuero @bwspenc

tophmatthews commented:
This seems seemingly simple! NICE! I'll profile tomorrow and see how it does.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/541246/) on 493d44f wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/15405/site)

This comment will be updated on new commits.

tophmatthews commented:
This shaved off 13% from one of my runs. NICE!!

dschwen commented on framework/include/utils/RankFourTensorImplementation.h:
```diff
@@ -542,62 +541,151 @@ RankFourTensorTempl<T>::fillFromInputVector(const std::vector<T> & input, FillMe
-  if ((all == true && input.size() != 21) || (all == false && input.size() != 9))
-    mooseError("Please check the number of entries in the stiffness input vector.");
+  if (input.size() != 9)
+    mooseError(
```
It would be appropriate to make this a `mooseAssert` (i.e. check only in debug mode).

lynnmunday replied:
I changed a few to asserts, should I have changed all of them?

dschwen commented on unit/src/RankFourTensorTest.C:
```diff
@@ -28,6 +28,32 @@ TEST(RankFourTensor, invSymm1)
+  }
+
+  RankFourTensor a(input, RankFourTensor::symmetric21);
+  EXPECT_NEAR(0, (iSymmetric - a.invSymm() * a).L2norm(), 1E-5);
```
No! This is not a test for the fill method at all. As discussed in the call you need to have a gold standard tensor (with _all_ 81 entries). Prepare that using the unmodified fill method. Then you compare if the new fill method produced the same result. 

Your test would pass for _any_ invertible tensor! It just tests `invSym` again. 

Same for the function below.

dschwen reviewed (changes requested):
Fix unit test, please. Make the unit test the _first_ commit _without_ the r4t fill methods modified, so we can easily verify that the before (old fill method) passes the test as well as the after (new fill method).

lynnmunday committed:
unit test for anistropic rank four tensors ref #15404

lynnmunday committed:
refactor fourth order tensor fill methods for efficiency ref #15404

-This hardcodes the filling of elastic materials properties fourth order tensors
-The fill method was also split into two methods depending on the type of anisotropy
-remove unnecessary zero'ing of tensor
-replace some if checks with asserts.

tophmatthews commented on framework/include/utils/RankFourTensorImplementation.h:
```diff
@@ -542,62 +541,151 @@ RankFourTensorTempl<T>::fillFromInputVector(const std::vector<T> & input, FillMe
 {
-  if ((all == true && input.size() != 21) || (all == false && input.size() != 9))
-    mooseError("Please check the number of entries in the stiffness input vector.");
+  if (input.size() != 9)
```
`mooseAssert(input.size() == 9, "Input must be size 9 to use fillSymmetric9FromInputVector)`. If statement is built in.

lynnmunday replied:
I fixed them.  Thanks for the reminder!

bwspenc commented:
This all happened when I was out, thanks for taking care of this!  I should leave more often!

@tophmatthews When you say that you see a 13% speedup, is that an AD run?  I'm curious what the speedup is for AD and non-AD.

tophmatthews commented:
13% in a AD run. Haven't tested for non-AD

lynnmunday commented:
> Fix unit test, please. Make the unit test the _first_ commit _without_ the r4t fill methods modified, so we can easily verify that the before (old fill method) passes the test as well as the after (new fill method).

This is done
